### PR TITLE
TXM In-memory: step 3-03-SaveFetchedReceipts

### PR DESCRIPTION
**Pull request by poopoothegorilla**
NOTES:
- Implements SaveFetchedReceipts

Parent:
- https://github.com/smartcontractkit/chainlink/pull/12179

### CODE REVIEW

File: common/txmgr/inmemory_store.go [hunk 1]
```diff
@@ -231,7 +233,37 @@ func (ms *inMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Updat
 }
 
 func (ms *inMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) SaveFetchedReceipts(ctx context.Context, receipts []R, chainID CHAIN_ID) error {
-	return nil
+	if ms.chainID.String() != chainID.String() {
+		return fmt.Errorf("save_fetched_receipts: %w: %s", ErrInvalidChainID, chainID)
```
Comment (amit-momin) [inline common/txmgr/inmemory_store.go:237]: Think we panic in other methods for this case

File: common/txmgr/inmemory_store.go [hunk 2]
```diff
@@ -231,7 +233,37 @@ func (ms *inMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Updat
...
+		for _, as := range ms.addressStates {
+			err := as.moveTxToConfirmed(receipt)
+			if err == nil {
+				found = true
+				break
+			}
+			receiptErrs = errors.Join(receiptErrs, err)
+		}
+		if !found {
+			receiptErrs = fmt.Errorf("receipt was not found: %w", receiptErrs)
```
Comment (amit-momin) [inline common/txmgr/inmemory_store.go:261]: Technically we didn't find the tx for the given receipt. Could you update this error message to reflect?

File: core/chains/evm/txmgr/evm_inmemory_store_test.go [hunk 3]
```diff
@@ -1,14 +1,112 @@
...
+	db := pgtest.NewSqlxDB(t)
+	_, dbcfg, evmcfg := evmtxmgr.MakeTestConfigs(t)
+	persistentStore := cltest.NewTestTxStore(t, db, dbcfg)
+	kst := cltest.NewKeyStore(t, db, dbcfg)
+	_, fromAddress := cltest.MustInsertRandomKey(t, kst.Eth())
+
+	ethClient := evmtest.NewEthClientMockWithDefaultChain(t)
+	lggr := logger.TestSugared(t)
+	chainID := ethClient.ConfiguredChainID()
+	ctx := context.Background()
```
Comment (amit-momin) [inline core/chains/evm/txmgr/evm_inmemory_store_test.go:40]: Should use `testutils.Context(t)` for tests here and below.

File: common/txmgr/address_state.go [hunk 4]
```diff
@@ -325,31 +325,37 @@ func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveIn
...
 	if !ok {
-		return fmt.Errorf("move_unconfirmed_to_confirmed: no unconfirmed transaction with ID %d", txAttempt.TxID)
+		return fmt.Errorf("no transaction with ID %d found for hash %q", txAttempt.TxID, receipt.GetTxHash())
 	}
 	txAttempt.State = txmgrtypes.TxAttemptBroadcast
 	if txAttempt.BroadcastBeforeBlockNum == nil {
 		blockNum := receipt.GetBlockNumber().Int64()
 		txAttempt.BroadcastBeforeBlockNum = &blockNum
 	}
 	tx.State = TxConfirmed
```
Comment (amit-momin) [inline common/txmgr/address_state.go:352]: If we're marking the tx as confirmed here, wouldn't we need to move it from the map that it's in to the confirmedTxs map?

Comment (DylanTinianov) [reply common/txmgr/address_state.go:352]: Yeah I think we really need to optimize everything into a single map to help avoid issues such as this. Made a ticket: https://smartcontract-it.atlassian.net/browse/BCI-2869. Either way we need to add a check here to move the state in the map.

File: common/txmgr/address_state.go [hunk 5]
```diff
@@ -350,6 +350,8 @@ func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveTx
...
+	as.confirmedTxs[tx.ID] = tx
```
Comment (amit-momin) [inline common/txmgr/address_state.go:353]: Just want to confirm here too that tx are only moved from Unconfirmed to Confirmed?

**Review by amit-momin — APPROVED** [review]

**Pull request merged**